server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -352,24 +355,62 @@ public void merge(IndexMetadata indexMetadata, MergeReason reason) {
         }
     }
 
+    /**
+     * Merging the provided mappings. If the provided mapping list is ordered, merging will maintain the original ordering.
+     */
+    public DocumentMapper merge(String type, List<CompressedXContent> mappingSources, MergeReason reason) {
+        final DocumentMapper currentMapper = this.mapper;
+        if (currentMapper != null && mappingSources.size() == 1 && currentMapper.mappingSource().equals(mappingSources.get(0))) {
+            return currentMapper;
+        }
+
+        List<Map<String, Object>> convertedMappings = new LinkedList<>();
+        Explicit<Boolean> explicitSubobjects = null;
+        for (CompressedXContent mappingSource : mappingSources) {
+            Map<String, Object> mapping = MappingParser.convertToMap(mappingSource);
+            convertedMappings.add(mapping);
+            Explicit<Boolean> tmpSubobjects = ObjectMapper.checkSubobjectsInMappingRoot(mapping);
+            if (tmpSubobjects != null) {
+                if (explicitSubobjects != null && tmpSubobjects.value() != explicitSubobjects.value()) {
+                    throw new MapperParsingException("Failed to parse mappings: contradicting subobjects settings provided");
+                }
+                explicitSubobjects = tmpSubobjects;
+            }
+        }
+
+        DocumentMapper ret = null;
+        for (Map<String, Object> mapping : convertedMappings) {
+            ret = doMerge(type, reason, mapping, explicitSubobjects);
+        }
+        return ret;
+    }
+
     public DocumentMapper merge(String type, CompressedXContent mappingSource, MergeReason reason) {
         final DocumentMapper currentMapper = this.mapper;
         if (currentMapper != null && currentMapper.mappingSource().equals(mappingSource)) {
             return currentMapper;
         }
-        synchronized (this) {
-            Mapping incomingMapping = parseMapping(type, mappingSource);
-            Mapping mapping = mergeMappings(this.mapper, incomingMapping, reason);
-            // TODO: In many cases the source here is equal to mappingSource so we need not serialize again.
-            // We should identify these cases reliably and save expensive serialization here
-            DocumentMapper newMapper = newDocumentMapper(mapping, reason, mapping.toCompressedXContent());
-            if (reason == MergeReason.MAPPING_UPDATE_PREFLIGHT) {
-                return newMapper;
-            }
-            this.mapper = newMapper;
-            assert assertSerialization(newMapper);
+        Map<String, Object> mappingSourceAsMap = MappingParser.convertToMap(mappingSource);
```

Comment:
Only the `XContentHelper.convertToMap()` operation was extracted from the formerly `synchronized` block. As far as I can see this is a stateless operation, which leaves the former synchronization enforcement the same.